trieste/utils/pareto.py

```diff
@@ -0,0 +1,309 @@
+from typing import Mapping
+
+import gpflow
+import numpy as np
+import tensorflow as tf
+
+from trieste.data import Dataset
+from trieste.type import TensorType
+
+jitter = gpflow.config.default_jitter()
+
+
+def atleast2d(tensor: tf.Tensor) -> tf.Tensor:
+    if tf.rank(tensor) <= 1:
+        return tf.expand_dims(tensor, axis=0)
+    return tensor
+
+
+def to_Dataset(datasets: Mapping[str, Dataset]) -> Dataset:
+
+    data_query_points = []
+    data_observations = []
+
+    for tag in datasets:
+        data_query_points.append(datasets[tag].query_points)
+        data_observations.append(datasets[tag].observations)
+
+    assert tf.math.reduce_all(
+        [tf.math.equal(input_data, data_query_points[0]) for input_data in data_query_points]
+    ), ValueError("Input data is not the same for each objective")
+
+    mo_dataset = Dataset(data_query_points[0], tf.concat(data_observations, axis=1))
+
+    return mo_dataset
+
+
+def non_dominated_sort(datasets: Dataset) -> TensorType:
+    """
+    Computes the non-dominated set for a set of data points
+    :param dataset: data points
+    :return: tuple of the non-dominated set and the degree of dominance,
+        dominances gives the number of dominating points for each data point
+    """
+    observations = datasets.observations
+    extended = tf.tile(tf.expand_dims(observations, 0), [observations.shape[0], 1, 1])
+    swapped_ext = tf.einsum("ij...->ji...", extended)
+    dominance = tf.reduce_sum(
+        tf.cast(
+            tf.logical_and(
+                tf.reduce_all(extended <= swapped_ext, axis=2),
+                tf.reduce_any(extended < swapped_ext, axis=2),
+            ),
+            tf.float32,
```

Comment:
why the cast?

Reply:
`tf.reduce_sum` cant work on boolean returned by `tf.logical_and`. should I use `tf.math.count_nonzero` or keep the cast and use gpflow default float?

Reply:
oh yes, of course. might an integer type be suitable?